File: lora-codec-adeunis/src/main/java/lora/codec/adeunis/AdeunisCodec.java
package lora.codec.adeunis;

import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import com.cumulocity.rest.representation.inventory.ManagedObjectRepresentation;
import com.google.common.collect.Lists;
import com.google.common.io.BaseEncoding;

import c8y.Configuration;
import c8y.RequiredAvailability;
import lora.codec.C8YData;
import lora.codec.DeviceCodec;
import lora.codec.DeviceOperation;
import lora.codec.DownlinkData;

@Component
public class AdeunisCodec extends DeviceCodec {

	private final Logger logger = LoggerFactory.getLogger(getClass());
	
	private Map<String, DeviceOperation> operations = new HashMap<String, DeviceOperation>();
	{
		operations.put("get config", new DeviceOperation("get config", "get config", null));
	}

	private ScriptEngine engine;
	{
		ScriptEngineManager manager = new ScriptEngineManager();
		engine = manager.getEngineByName("nashorn");
		try {
			engine.eval(new InputStreamReader(getClass().getResourceAsStream("/buffer-custom-shim.js")));
			engine.eval(new InputStreamReader(getClass().getResourceAsStream("/lib.js")));
			engine.eval("var decoder = new codec.Decoder();");
			engine.eval("var encoder = new codec.Encoder();");
		} catch (ScriptException e) {
			e.printStackTrace();
		}
	}

	@Override
	public String getId() {
		return "adeunis";
	}

	@Override
	public String getName() {
		return "Adeunis";
	}

	@Override
	public String getVersion() {
		return "1.0";
	}

	@Override
	protected C8YData decode(ManagedObjectRepresentation mor, String model, int fport, DateTime updateTime, byte[] payload) {
		C8YData c8yData = new C8YData();
		
		try {
			engine.eval("decoder.setDeviceType('" + model + "');");
			engine.eval("var result = decoder.decode('" + BaseEncoding.base16().encode(payload) +"');");
			engine.eval("var payloadResult = JSON.stringify(result, null, 2);");
			Map result = (Map)engine.get("result");
			String payloadResult = engine.get("payloadResult").toString();
			logger.info(payloadResult);
			if (result.containsKey("temperature")) {
				String unit = ((Map)result.get("temperature")).get("unit").toString();
				Object intermediateValue = ((Map)((Map)result.get("temperature")).get("values")).get("0");
				BigDecimal value = intermediateValue instanceof Double ? BigDecimal.valueOf((double)intermediateValue) : intermediateValue instanceof Integer ? BigDecimal.valueOf((int)intermediateValue) : null;
				String series = "T";
				c8yData.addMeasurement(mor, "c8y_TemperatureMeasurement", series, unit, value, new DateTime());
			}
			if (result.containsKey("temperatures")) {
				Map<String, Map> temperatures = (Map)result.get("temperatures");
				for (Map temperature: temperatures.values()) {
					String unit = temperature.get("unit").toString();
					Object intermediateValue = temperature.get("value");
					BigDecimal value = intermediateValue instanceof Double ? BigDecimal.valueOf((double)intermediateValue) : intermediateValue instanceof Integer ? BigDecimal.valueOf((int)intermediateValue) : null;
					String series = temperature.get("name").toString();
					c8yData.addMeasurement(mor, "Temperature", series, unit, value, new DateTime());
				}
			}
			if (result.containsKey("humidity")) {
				String unit = ((Map)result.get("humidity")).get("unit").toString();
				Object intermediateValue = ((Map)((Map)result.get("humidity")).get("values")).get("0");
				BigDecimal value = intermediateValue instanceof Double ? BigDecimal.valueOf((double)intermediateValue) : intermediateValue instanceof Integer ? BigDecimal.valueOf((int)intermediateValue) : null;
				String series = "H";
				c8yData.addMeasurement(mor, "Humidity", series, unit, value, new DateTime());
			}
			if (result.containsKey("type")) {
				String type = result.get("type").toString();
				if (type.contains("configuration")) {
					mor.set(new Configuration(payloadResult));
					if (result.containsKey("calculatedSendingPeriod")) {
						String unit = ((Map)result.get("calculatedSendingPeriod")).get("unit").toString();
						double requiredAvailability = (double)((Map)result.get("calculatedSendingPeriod")).get("value");
						if (unit.equals("s")) {
							requiredAvailability = requiredAvailability / 60.0;
						}
						mor.set(new RequiredAvailability((int)requiredAvailability));
					}
				}
			}
		} catch (ScriptException e) {
			e.printStackTrace();
		}

		
		return c8yData;
	}

	@Override
	public List<String> getModels() {
		List<String> models = new ArrayList<String>();
		try {
			engine.eval("var intermediate = codec.DecoderProducts;");
			Map intermediate = (Map)engine.get("intermediate");
			models = Lists.newArrayList(intermediate.keySet());
		} catch (ScriptException e) {
			e.printStackTrace();
		}
		return models;
	}

	@Override
	protected DownlinkData encode(ManagedObjectRepresentation mor, String model, String operation) {
		/*DownlinkData data = new DownlinkData();
		try {
			engine.eval("encoder.setDeviceType('" + model + "');");
			engine.eval("var result = encoder.encode('" + model + "', '" + operation + "');");
			engine.eval("var payloadResult = JSON.stringify(result, null, 2);");
			String result = (String)engine.get("payloadResult");
			data.setPayload(result);
		} catch (ScriptException e) {
			e.printStackTrace();
		}
		return data;*/
		return null;
	}

	@Override
	public DownlinkData askDeviceConfig(String devEui) {
		return new DownlinkData(devEui, 1, "01");
	}

	@Override
	public Map<String, DeviceOperation> getAvailableOperations(String model) {
		return operations;
	}

}

File: nodejs/lora-codec-adeunis/src/AdeunisCodec.ts
import { Client, IManagedObject, Severity } from "@c8y/client";
import {
  C8YData,
  DeviceCodec,
  DeviceOperation,
  DownlinkData,
} from "lora-codec-interface";
const codec = require("@adeunis/codecs");
require("source-map-support").install();

export class AdeunisCodec extends DeviceCodec {
  private encoder = new codec.Encoder();
  private decoder = new codec.Decoder();

  getId(): string {
    return "adeunis";
  }
  getName(): string {
    return "Adeunis";
  }
  getVersion(): string {
    return "1.0";
  }
  getModels(client: Client): Map<string, string> {
    let models: Map<string, string> = new Map<string, string>();
    for (let model in codec.DecoderProducts) {
      models.set(model, model);
    }
    return models;
  }
  askDeviceConfig(devEui: string): DownlinkData {
    return { devEui: devEui, fport: 1, payload: "01" };
  }
  getAvailableOperations(
    client: Client,
    model: string
  ): Map<string, DeviceOperation> {
    let operations: Map<string, DeviceOperation> = new Map<
      string,
      DeviceOperation
    >();
    operations.set("get config", {
      id: "get config",
      name: "get config",
      elements: [],
    });
    return operations;
  }
  protected _decode(
    client: Client,
    mo: Partial<IManagedObject>,
    model: string,
    fport: number,
    time: Date,
    payload: string
  ): C8YData {
    let c8yData: C8YData = new C8YData();
    this.decoder.setDeviceType(model);
    let result = this.decoder.decode(payload);
    console.log(result);

    this.handleStatusByte(result.status, mo, time, c8yData);

    if (result.temperatures) {
      result.temperatures.forEach((t) => {
        c8yData.addMeasurement(
          mo,
          "Temperatures",
          t.name,
          t.unit,
          t.value,
          time
        );
      });
    }
    if (result.temperature) {
      c8yData.addMeasurement(
        mo,
        "c8y_TemperatureMeasurement",
        "T",
        result.temperature.unit,
        result.temperature.values[0],
        time
      );
    }
    if (result.humidity) {
      c8yData.addMeasurement(
        mo,
        "Humidity",
        "H",
        result.humidity.unit,
        result.humidity.values[0],
        time
      );
    }
    if (
      (result.type?.includes("0x5a") || result.type?.includes("0x5b")) &&
      result.counterValues
    ) {
      if (!mo.samplingPeriod) {
        console.error("Device must first retrieve its configuration.");
      }
      let samplingPeriod = mo.samplingPeriod.value;
      if (mo.samplingPeriod.unit === "m") {
        samplingPeriod = samplingPeriod * 60;
      }
      console.log("Sampling period: " + samplingPeriod);
      let channel = "Channel A";
      if (result.type?.includes("0x5b")) {
        channel = "Channel B";
      }
      result.counterValues.forEach((c, i) => {
        let datetime = new Date(time.getTime() - samplingPeriod * i * 1000);
        console.log("Time: " + datetime);
        c8yData.addMeasurement(
          mo,
          "Pulse",
          channel,
          "",
          c,
          new Date(time.getTime() - samplingPeriod * i * 1000)
        );
      });
    }
    if (result.type?.includes("configuration")) {
      mo["c8y_Configuration"] = { config: JSON.stringify(result) };
      if (result.samplingPeriod) {
        mo.samplingPeriod = result.samplingPeriod;
      }
      if (result.calculatedSendingPeriod) {
        let requiredAvailability: number = result.calculatedSendingPeriod.value;
        if (result.calculatedSendingPeriod.unit === "s") {
          requiredAvailability = requiredAvailability / 60.0;
        }
        mo["c8y_RequiredAvailability"] = {
          responseInterval: requiredAvailability,
        };
      }
      if (result.transmissionPeriod) {
        let requiredAvailability: number = result.transmissionPeriod.value;
        if (result.transmissionPeriod.unit === "s") {
          requiredAvailability = requiredAvailability / 60.0;
        }
        mo["c8y_RequiredAvailability"] = {
          responseInterval: requiredAvailability,
        };
      }
      c8yData.morToUpdate = mo;
    }
    // Analog
    if (result.type?.includes("0x42") && result.channels) {
      result.channels.forEach((channel) => {
        c8yData.addMeasurement(
          mo,
          channel.name,
          channel.name,
          channel.unit,
          channel.value,
          time
        );
      });
    }
    // Pulse historic data
    if (result.type?.includes("0x48") && result.channels) {
      result.channels.forEach((channel) => {
        channel.deltaValues.forEach((value, i) => {
          c8yData.addMeasurement(
            mo,
            channel.name,
            channel.name,
            "count",
            channel.index + value,
            new Date(time.getTime() + result.baseTime * i * 60000)
          );
        });
      });
    }
    // Pulse data
    if (result.type?.includes("0x46") && result.counterValues) {
      result.counterValues.forEach((value, i) => {
        c8yData.addMeasurement(
          mo,
          "Channel " + String.fromCharCode(65 + i),
          "Channel " + String.fromCharCode(65 + i),
          "count",
          value,
          time
        );
      });
    }
    return c8yData;
  }
  protected _encode(
    client: Client,
    mo: Partial<IManagedObject>,
    model: string,
    operation: string
  ): DownlinkData {
    return operation.includes("get config") ? this.askDeviceConfig(null) : null;
  }

  private handleStatusByte(
    status,
    mo: Partial<IManagedObject>,
    time: Date,
    c8yData: C8YData
  ) {
    if (status.hardwareError) {
      c8yData.addAlarm(
        mo,
        "HardwareErrorAlarm",
        "Hardware error",
        Severity.CRITICAL,
        time
      );
    } else {
      c8yData.clearAlarm("HardwareErrorAlarm");
    }

    if (status.lowBattery) {
      c8yData.addAlarm(
        mo,
        "LowBatteryAlarm",
        "Battery is low",
        Severity.MAJOR,
        time
      );
    } else {
      c8yData.clearAlarm("LowBatteryAlarm");
    }

    if (status.configurationDone) {
      c8yData.addEvent(
        mo,
        "ConfigurationDone",
        "Configuration is done",
        null,
        time
      );
    }

    if (status.configurationInconsistency) {
      c8yData.addAlarm(
        mo,
        "InconsistentConfigurationAlarm",
        "Configuration is inconsistent",
        Severity.MAJOR,
        time
      );
    } else {
      c8yData.clearAlarm("InconsistentConfigurationAlarm");
    }
  }
}
